### PR TITLE
SSS: value-set-analysis adaptation

### PR DESCRIPTION
**Pull request by smowton**
This makes a couple of small additions to VSA (storing function-ID in each value set, and a declared-type in value set entries), and templatizes and virtualizes it to make it suitable for subclassing and adaptation by the security scanner.

### CODE REVIEW

File: src/pointer-analysis/value_set.cpp [hunk 1]
```diff
@@ -1062,6 +1062,32 @@ void value_sett::get_reference_set(
     dest.push_back(to_expr(it));
 }
 
+static void strip_casts(
+  exprt& e,
```
Comment (peterschrammel) [inline src/pointer-analysis/value_set.cpp]: &

**Comment by smowton** [comment]
Done

File: src/pointer-analysis/value_set.cpp [hunk 2]
```diff
@@ -1062,6 +1062,32 @@ void value_sett::get_reference_set(
     dest.push_back(to_expr(it));
 }
 
+static void strip_casts(
```
Comment (peterschrammel) [inline src/pointer-analysis/value_set.cpp]: function comment block missing

File: src/pointer-analysis/value_set.cpp [hunk 3]
```diff
@@ -1062,6 +1062,32 @@ void value_sett::get_reference_set(
...
+    {
+      auto &mem=to_member_expr(e);
+      const auto &struct_type=to_struct_type(ns.follow(e.op0().type()));
+      if(mem.get_component_name()==struct_type.components()[0].get_name())
+        e=e.op0();
+      else
+        return;
+    }
+    else
+      return;
```
Comment (peterschrammel) [inline src/pointer-analysis/value_set.cpp]: maybe add a blank line after the return

File: src/pointer-analysis/value_set_analysis.h [hunk 4]
```diff
@@ -11,32 +11,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+      if(location.is_nil() || location.get_file()==irep_idt())
+        continue;
+
+      // find value set
+      const value_sett &value_set=(*this)[i_it].value_set;
+
+      xmlt &i=dest.new_element("instruction");
+      i.new_element()=::xml(location);
+
+      for(value_sett::valuest::const_iterator
```
Comment (peterschrammel) [inline src/pointer-analysis/value_set_analysis.h]: for each

File: src/pointer-analysis/value_set_analysis.h [hunk 5]
```diff
@@ -11,32 +11,90 @@ Author: Daniel Kroening, kroening@kroening.com
...
+    source_locationt previous_location;
+
+    forall_goto_program_instructions(i_it, goto_program)
+    {
+      const source_locationt &location=i_it->source_location;
+
+      if(location==previous_location)
+        continue;
+
+      if(location.is_nil() || location.get_file()==irep_idt())
```
Comment (peterschrammel) [inline src/pointer-analysis/value_set_analysis.h]: get_file().empty()

**Comment by smowton** [comment]
Made first two changes. The latter two are inherited from existing VSA, so I won't make them here (they will only clash with likely-upcoming upstream changes)

**Pull request merged**